### ISSUE TITLE
Expose outgoing arguments in ICompletedFakeObjectCall

### ISSUE THREAD

**Issue by thomaslevesque** [FakeItEasy/FakeItEasy#1140]
Related to #1124 and #1138

From https://github.com/FakeItEasy/FakeItEasy/pull/1138#discussion_r122580359:

> [@blairconrad] It's a little troubling that Arguments means two different things, depending on the type of the underlying object implementing IFakeObjectCall. I'm not sure what to do about it, though.

> [@thomaslevesque] True. Maybe we should change the API in 4.0.0 to expose both.

**Comment by blairconrad** [comment]
Potentially breaking.

**Comment by thomaslevesque** [comment]
> Potentially breaking.

Technically breaking, since it would modify a public interface.

**Comment by blairconrad** [comment]
It's true, but it's one of those that there's no reason for users to modify, so I'm less worked up. Not worked up enough to treat it as non-breaking, though.

**Comment by thomaslevesque** [comment]
If we want to do this, it would be nice to do it in 4.0.0, otherwise it will have to wait until the next major.

**Comment by thomaslevesque** [comment]
Suggested changes to `IFakeObjectCall`/`ICompletedFakeObjectCall`:

```csharp
    /// <summary>
    /// Represents a call to a fake object at interception time.
    /// </summary>
    public interface IFakeObjectCall
    {
        /// <summary>
        /// Gets the method that's called.
        /// </summary>
        MethodInfo Method { get; }

        /// <summary>
        /// Gets the arguments used in the call, before the call is made.
        /// </summary>
        ArgumentCollection ArgumentsBeforeCall { get; }

        /// <summary>
        /// Gets the faked object the call is performed on.
        /// </summary>
        object FakedObject { get; }
    }

    /// <summary>
    /// Represents a completed call to a fake object.
    /// </summary>
    public interface ICompletedFakeObjectCall : IFakeObjectCall
    {
        /// <summary>
        /// Gets the value set to be returned from the call.
        /// </summary>
        object ReturnValue { get; }
        /// <summary>
        /// Gets the arguments used in the call, after the call is made. Includes changes to the values of out and ref arguments.
        /// </summary>
        ArgumentCollection ArgumentsAfterCall { get; }
    }
```

Note that renaming `Arguments` to `ArgumentsBeforeCall` is a significant breaking change, because all code that uses it (e.g. in `.Where(...)`) will have to be updated. We could also *not* rename it, just clarify its meaning and introduce the new `ArgumentsAfterCall` property.

**Comment by blairconrad** [comment]
@thomaslevesque, I generally like the idea, but would shy away from renaming `Arguments`.  As you suggest, the change is too great. I think your idea to not rename and to introduce the new property is a better plan.

I'd been working on an analysis of `IFakeObjectCall`, `ICompletedFakeObjectCall` and `IInterceptedFakeObjectCall` before I went away, and I'd like to complete it before we release 4.0.0, but I don't see any reason not to proceed as you've outlined in the meantime.

**Comment by blairconrad** [comment]
Okay, yeah. My analysis wasn't as deep as I though it was. I was worried about `IInterceptedFakeObjectCall.SetArgumentValue`. If there's a new implementation of `IInterceptedFakeObjectCall`, it'll have to play nicely with the `Arguments`/`ArgumentsAfterCall` values, but I'm content that we won't have to deal with a user-supplied implementation of `IInterceptedFakeObjectCall`.

**Comment by thomaslevesque** [comment]
> I was worried about `IInterceptedFakeObjectCall.SetArgumentValue`.

Me too! I even wondered briefly if we should remove `IInterceptedFakeObjectCall` from the public API, but there are legitimate use-cases for it (custom call rules, although I've never seen anyone use them)

> If there's a new implementation of `IInterceptedFakeObjectCall`, it'll have to play nicely with the `Arguments`/`ArgumentsAfterCall` values, but I'm content that we won't have to deal with a user-supplied implementation of `IInterceptedFakeObjectCall`.

Sounds reasonable.

I wonder if `IInterceptedFakeObjectCall` should also expose `ArgumentsAfterCall` and `ReturnValue`. It's a bit strange to be able to set those values, but not read them back. BTW, I think the fix for #1124 introduced a breaking change. Before that, it was possible to read the values set by `SetArgumentValue`, but it's no longer the case. I doubt anyone will be affected, though.

**Comment by blairconrad** [comment]
Yeah, the #1124 would totally break that use case. Sorry, I thought it was understood. My interpretation was that we were intending to expose the pre-call arguments, and so our change fixed that bug, so the use of those values was a "relying on a bug" case. And now exposing the post-call arguments is an enhancement!

> I wonder if `IInterceptedFakeObjectCall` should also expose `ArgumentsAfterCall` and `ReturnValue`

Are they necessarily available when user code sees an `IInterceptedFakeObjectCall`? I'm away from the code right now. I'll check later. At this point, I'm not terribly worried about being able to set a value and not read it back. Will get back to this later.

**Comment by blairconrad** [comment]
@thomaslevesque, `IInterceptedFakeObjectCall` is user-facing only during the execution of `IFakeObjectCallRule.Apply`. If `ArgumentsAfterCall` and `ReturnValue` are available at this time, it's because the rule is setting them. I don't think we need to expose them.

**Comment by thomaslevesque** [comment]
> @thomaslevesque, IInterceptedFakeObjectCall is user-facing only during the execution of IFakeObjectCallRule.Apply. If ArgumentsAfterCall and ReturnValue are available at this time, it's because the rule is setting them. I don't think we need to expose them.

Good point 👍

**Comment by blairconrad** [comment]
Are you fond of this issue, or shall I take it? You've done the hard work of proposing an interface, after all…

**Comment by thomaslevesque** [comment]
> Are you fond of this issue, or shall I take it?

Not especially fond of it, please go ahead ;)

**Comment by blairconrad** [comment]
This change has been released as part of [FakeItEasy 4.0.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/4.0.0).
